modules/fedimint-mint-client/src/input.rs

```diff
@@ -161,6 +176,104 @@ impl MintInputStateCreated {
+                    dbtx,
+                    global_context.clone(),
+                ))
+            },
+        )]
+    }
+
+    async fn await_success(
+        common: MintInputCommon,
+        global_context: DynGlobalClientContext,
+    ) -> Result<(), String> {
+        global_context.await_tx_accepted(common.txid).await
+    }
+
+    async fn transition_success(
+        result: Result<(), String>,
+        old_state: MintInputStateMachine,
+        dbtx: &mut ClientSMDatabaseTransaction<'_, '_>,
+        global_context: DynGlobalClientContext,
+    ) -> MintInputStateMachine {
+        assert!(matches!(old_state.state, MintInputStates::CreatedMulti(_)));
+
+        match result {
+            Ok(()) => {
+                // Success case: containing transaction is accepted
+                MintInputStateMachine {
+                    common: old_state.common,
+                    state: MintInputStates::Success(MintInputStateSuccess {}),
+                }
+            }
+            Err(err) => {
+                // Transaction rejected: attempting to refund
+                debug!(target: LOG_CLIENT_MODULE_MINT, %err, "Refunding mint transaction input due to transaction error");
+                Self::refund(dbtx, old_state, global_context).await
+            }
+        }
+    }
+
+    async fn refund(
+        dbtx: &mut ClientSMDatabaseTransaction<'_, '_>,
+        old_state: MintInputStateMachine,
+        global_context: DynGlobalClientContext,
+    ) -> MintInputStateMachine {
+        let spendable_notes = match old_state.state {
+            MintInputStates::CreatedMulti(created) => created.notes,
+            _ => panic!("Invalid state transition"),
+        };
+
+        let mut refund_txids = vec![];
+        for (amount, spendable_note) in spendable_notes {
+            let refund_input = ClientInput::<MintInput> {
+                input: MintInput::new_v0(amount, spendable_note.note()),
+                keys: vec![spendable_note.spend_key],
+                amount,
+            };
+            let (refund_txid, _) = global_context
+                .claim_inputs(
+                    dbtx,
+                    // The input of the refund tx is managed by this state machine, so no new state
+                    // machines need to be created
+                    ClientInputBundle::new_no_sm(vec![refund_input]),
+                )
+                .await
+                .expect("Cannot claim input, additional funding needed");
+
+            refund_txids.push(refund_txid);
+        }
+
+        MintInputStateMachine {
+            common: old_state.common,
+            state: match refund_txids.len() {
+                0 => panic!("Can't have nothing here"),
+                1 => MintInputStates::Refund(MintInputStateRefund {
+                    refund_txid: refund_txids[0],
+                }),
+                _ => MintInputStates::RefundMulti(MintInputStateRefundMulti { refund_txids }),
+            },
+        }
```

Comment:
Why not always make it a refund-multi from now on? We could even migrate old states into their multi equivalents this way, possibly avoiding having to maintain two versions of the update stream code.